lib/Dialect/AIG/Transforms/GreedyCutDecomp.cpp

```diff
@@ -0,0 +1,288 @@
+// Conversion patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct AndInverterOpToCutPattern : public OpRewritePattern<aig::AndInverterOp> {
+  using OpRewritePattern<aig::AndInverterOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(aig::AndInverterOp op,
+                                PatternRewriter &rewriter) const override {
+    if (isa<aig::CutOp>(op->getParentOp()))
+      return failure();
+
+    auto cutOp = rewriter.create<aig::CutOp>(
+        op.getLoc(), op.getResult().getType(), op.getInputs(), [&]() {
+          auto result = rewriter.create<aig::AndInverterOp>(
+              op.getLoc(), op.getResult().getType(),
+              rewriter.getBlock()->getArguments(), op.getInvertedAttr());
+          rewriter.create<aig::OutputOp>(op.getLoc(), ValueRange{result});
+        });
+
+    rewriter.replaceOp(op, cutOp);
+    return success();
+  }
+};
+
+static aig::CutOp mergeCuts(Location loc, MutableArrayRef<Operation *> cuts,
+                            ArrayRef<Value> inputs, Value output,
+                            PatternRewriter &rewriter) {
+  if (!mlir::computeTopologicalSorting(cuts))
+    return {};
+
+  assert(cuts.size() >= 2);
+
+  DenseMap<Value, Value> valueToNewValue, inputsToBlockArg;
+  auto cutOp =
+      rewriter.create<aig::CutOp>(loc, output.getType(), inputs, [&]() {
+        for (auto [i, input] : llvm::enumerate(inputs))
+          inputsToBlockArg[input] = rewriter.getBlock()->getArgument(i);
+
+        for (auto [i, cut] : llvm::enumerate(cuts)) {
+          auto cutOp = cast<aig::CutOp>(cut);
+          assert(cutOp.getNumResults() == 1);
+          for (auto [arg, input] : llvm::zip(
+                   cutOp.getBodyBlock()->getArguments(), cutOp.getInputs())) {
+            auto it = inputsToBlockArg.find(input);
+            if (it != inputsToBlockArg.end()) {
+              arg.replaceAllUsesWith(it->second);
+            } else {
+              auto cutOp = dyn_cast<aig::CutOp>(input.getDefiningOp());
+              assert(cutOp && cutOp.getNumResults() == 1);
+              arg.replaceAllUsesWith(valueToNewValue.at(input));
+            }
+          }
+
+          assert(cutOp.getNumResults() == 1);
+          valueToNewValue[cutOp.getResult(0)] =
+              cutOp.getBodyBlock()->getTerminator()->getOperand(0);
+        }
+      });
+
+  output.replaceAllUsesWith(cutOp.getResult(0));
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToEnd(cutOp.getBodyBlock());
+    rewriter.create<aig::OutputOp>(loc, ValueRange{valueToNewValue.at(output)});
+  }
+
+  for (auto oldCut : llvm::reverse(cuts)) {
+    auto oldCutOutput =
+        cast<aig::CutOp>(oldCut).getBodyBlock()->getTerminator();
+    oldCutOutput->erase();
+    cutOp.getBodyBlock()->getOperations().splice(
+        cutOp.getBodyBlock()->begin(),
+        cast<aig::CutOp>(oldCut).getBodyBlock()->getOperations());
```

Comment:
To be safe, should this mutation and the above use the rewriter?

Reply:
Yeah that's great point. I'll modify it through rewriter.